File: Archive_LJ_grand_canonical/zgcmclj.c
/***********************************************************************
 *
 * File: zgcmclj.c
 *
 * Grand Canonical (TVmu) Monte Carlo of Lennard-Jonesium
 * Re-initialize checkpoint file
 *
 * 08-Apr-2010 (MN)
 * 18-Apr-2012
 *
 ***********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define BSIZE 80

/***********************************************************************/

int main() {

/***********************************************************************/

  FILE *fpi,*fpo;
  char fname[BSIZE];
  unsigned short seed[3];
  int n,nt,ntjob,ntprint,ntskip;
  int i;
  double disp,dr,pcre,t,v,zz;
  double fact,v_old;
  double *x,*y,*z;

  /* Read old checkpoint file */

  strcpy(fname,"gcmclj_out.dat");
  printf("         infile=[gcmclj_out.dat] ");
  getval_s(fname);

  fpi=fopen(fname,"r");

  fread(&n,sizeof(int),1,fpi);
  fread(&nt,sizeof(int),1,fpi);
  fread(&ntjob,sizeof(int),1,fpi);
  fread(&ntprint,sizeof(int),1,fpi);
  fread(&ntskip,sizeof(int),1,fpi);
  fread(seed,sizeof(unsigned short),3,fpi);

  fread(&disp,sizeof(double),1,fpi);
  fread(&dr,sizeof(double),1,fpi);
  fread(&pcre,sizeof(double),1,fpi);
  fread(&t,sizeof(double),1,fpi);
  fread(&v,sizeof(double),1,fpi);
  fread(&zz,sizeof(double),1,fpi);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Positions */

  fread(x,sizeof(double),n,fpi);
  fread(y,sizeof(double),n,fpi);
  fread(z,sizeof(double),n,fpi);

  fclose(fpi);

  /* User input */

  printf("              t=[%14.7lf] ",t);
  getval_d(&t);
  v_old=v;
  printf("              v=[%14.7lf] ",v);
  getval_d(&v);
  printf("              n=%15d\n",n);
  printf("              z=[%14.7lf] ",zz);
  getval_d(&zz);

  for(;;) {
    printf("     p(cre/del)=[%14.7lf] ",pcre);
    getval_d(&pcre);
    if(pcre<0.5) {
      break;
    }
  }

  printf("           disp=[%14.7lf] ",disp);
  getval_d(&disp);
  printf("             dr=[%14.7lf] ",dr);
  getval_d(&dr);
  printf("         ntskip=[%14d] ",ntskip);
  getval_i(&ntskip);
  printf(" ntprint/ntskip=[%14d] ",ntprint);
  getval_i(&ntprint);
  printf("   ntjob/ntskip=[%14d] ",ntjob);
  getval_i(&ntjob);

  strcpy(fname,"gcmclj_in.dat");
  printf("        outfile=[ gcmclj_in.dat] ");
  getval_s(fname);

  /* Rescale positions */

  if(v!=v_old) {
    fact=cbrt(v/v_old);
    for(i=0;i<=n-1;i++) {
      x[i]=fact*x[i];
      y[i]=fact*y[i];
      z[i]=fact*z[i];
    }
  }

  /* Write new startup file */

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&pcre,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);
  fwrite(&v,sizeof(double),1,fpo);
  fwrite(&zz,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/***********************************************************************/

File: Archive_LJ_grand_canonical/zgcmclj.f90
!**********************************************************************!
!
! File: zgcmclj.f90
!
! Grand Canonical (TVmu) Monte Carlo of Lennard-Jonesium
! Re-initialize startup/checkpoint file
!
! 08-Feb-2004 (MN)
! 18-Apr-2012
!
!**********************************************************************!

module getval_m

!**********************************************************************!

! Generic subroutine: read item from standard input/keep old value on
! hitting return

  implicit none

  private::getint,getreal,getstrg
  public::getval

  interface getval
    module procedure getint,getreal,getstrg
  end interface

  character(len=80),private::line

contains

!**********************************************************************!

  subroutine getint(intval)

!**********************************************************************!

    integer,intent(in out)::intval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) intval
    end if

    return

  end subroutine getint

!**********************************************************************!

  subroutine getreal(realval)

!**********************************************************************!

    real,intent(in out)::realval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) realval
    end if

    return

  end subroutine getreal

!**********************************************************************!

  subroutine getstrg(strgval)

!**********************************************************************!

    character(len=*),intent(in out)::strgval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt="(a)") strgval
    end if

    return

  end subroutine getstrg

end module getval_m

!**********************************************************************!

program zgcmclj

!**********************************************************************!

  use getval_m

  implicit none

  integer,parameter::iin=1,iout=2

  character(len=80)::fname
  integer::n,nt,ntjob,ntprint,ntskip
  integer::nran
  integer,dimension(:),allocatable::iran
  real::disp,dr,pcre,t,v,zz
  real::fact,v_old
  real,dimension(:),allocatable::x,y,z

! RNG characteristics

  call random_seed(size=nran)

  allocate(iran(nran))

! Read old checkpoint file

  fname="gcmclj_out.dat"
  write(unit=*,fmt="(a)",advance="no") &
       "         infile=[gcmclj_out.dat] "
  call getval(fname)

  open(unit=iin,file=fname,status="old",action="read", &
       form="unformatted",position="rewind")

  read(unit=iin) n,nt,ntjob,ntprint,ntskip,iran
  read(unit=iin) disp,dr,pcre,t,v,zz

! Allocate arrays

  allocate(x(n),y(n),z(n))

! Positions

  read(unit=iin) x,y,z

  close(unit=iin)

! User input

  write(unit=*,fmt="(a,f14.7,a)",advance="no") &
       "              t=[",t,"] "
  call getval(t)
  v_old=v
  write(unit=*,fmt="(a,f14.7,a)",advance="no") &
       "              v=[",v,"] "
  call getval(v)
  write(unit=*,fmt="(a,i15)") "              n=",n
  write(unit=*,fmt="(a,f14.7,a)",advance="no") &
       "              z=[",zz,"] "
  call getval(zz)

  do
    write(unit=*,fmt="(a,f14.7,a)",advance="no") &
         "     p(cre/del)=[",pcre,"] "
    call getval(pcre)
    if(pcre<0.5) then
      exit
    end if
  end do

  write(unit=*,fmt="(a,f14.7,a)",advance="no") &
       "           disp=[",disp,"] "
  call getval(disp)
  write(unit=*,fmt="(a,f14.7,a)",advance="no") &
       "             dr=[",dr,"] "
  call getval(dr)
  write(unit=*,fmt="(a,i14,a)",advance="no") &
       "         ntskip=[",ntskip,"] "
  call getval(ntskip)
  write(unit=*,fmt="(a,i14,a)",advance="no") &
       " ntprint/ntskip=[",ntprint,"] "
  call getval(ntprint)
  write(unit=*,fmt="(a,i14,a)",advance="no") &
       "   ntjob/ntskip=[",ntjob,"] "
  call getval(ntjob)

  fname="gcmclj_in.dat"
  write(unit=*,fmt="(a)",advance="no") &
       "        outfile=[ gcmclj_in.dat] "
  call getval(fname)

! Rescale positions

  if(v/=v_old) then
    fact=(v/v_old)**(1.0/3.0)
    x=fact*x
    y=fact*y
    z=fact*z
  end if

! Write new startup file

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dr,pcre,t,v,zz
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program zgcmclj

!**********************************************************************!
